posts/inside-rust/2023-09-29-polonius-update.md

```diff
@@ -0,0 +1,152 @@
+
+It's expected that performance will be similar, and we can then imagine enabling the location-insensitive pass without the feature flag, and removing some of the old code. 
+
+To keep the high quality diagnostics from the years of work of many contributors, it's possible that the new analysis could run, and if errors are detected, only then use the existing analysis and diagnostics.
+
+Current status: we've done some early investigations on the data-structures changes needed, some of redundant parts that could be removed, etc.
+
+### 5. Location-sensitive pass on nightly
+
+Then the (harder-to-do-efficiently) work to incorporate location-sensitivity can start. This step will implement the first version of the analysis.
+
+At this point it can still be inefficient, and use the feature flag, but this is when the borrow checker should accept more expressive code than the current NLLs.
+
+Current status: we're in the design phase here, to adapt our datalog prototype and algorithms to rustc, imagining alternative ways to compute and propagate the subset constraints along the CFG.
+
+### 6. Model borrow checking and polonius in a-mir-formality
+
+The types team is building a model of Rust's MIR and trait system called [`a-mir-formality`][]. Once it reaches a sufficiently complete status, the intent is that the model will always be extended to cover new language features prior to stabilization. We are therefore working to add polonius into the model. This will in fact be the second time doing such modeling, as we already added polonius to a previous incarnation of a-mir-formality. In fact, that modeling work is what gave us the insights that enabled the location-insensitive polonius formulation now landing on nightly.
+
+[`a-mir-formality`]: https://github.com/rust-lang/a-mir-formality
+
+Interestingly, this work is completely independent of rustc, and could in theory be started soon, and done in parallel with the other efforts.
+
+### 7. Location-sensitive pass stable
+
+In this milestone, we expect a lot of work on optimizations, and productization. 
+
+If a similar experience to NLLs in edition 2018 is to be expected again, another substantial amount of work and polish will also be needed to handle diagnostic differences and issues, ensuring errors and notes are clear enough, as well as the documentation.
+
+At this point, the location-sensitive pass is hopefully efficient enough, tested in practice, somewhat formally verified, and can be enabled in edition 2024.
+
+Around this time, librarification efforts can also be rebooted, to turn the in-tree polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][]), [prusti][] and [creusot][]).
```

Comment:
```suggestion
Around this time, librarification efforts can also be rebooted, to turn the in-tree Polonius into a library, maybe using [Stable MIR][]. This is so that it could be reused elsewhere, for example in [rust-analyzer][], or [gccrs][], or by researchers working on verification tools (like [kani][]), [prusti][] and [creusot][]).
```

Reply:
Also, missing links here.

Reply:
I've pushed a commit for the capitalization. The links are just below the sentence though.